[DOC] Adding guides to explain UDF serialization and Broadcast variable usage (#464)
dotnet/spark: docs/udf-guide.md

> # Guide to User-Defined Functions (UDF)

Reviewer: nit

```suggestion
# Guide to User-Defined Functions (UDFs)
```

This is a guide to show how to use UDFs in .NET for Apache Spark.

## What are UDFs

> [User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the vocabulary of Spark SQL’s Domain Specific Language. They transform values from a single row within a table to produce a single corresponding output value per row based on the logic defined in the UDF.

Reviewer: What is Spark SQL's Domain Specific Language? Could we provide a hyperlink and/or explanation to give some more context?

Author: Yeah on retrospect I think that is a jargon that doesn't need to be thrown at the user, modifying the sentence to make it more readable.

> Let's take the following as an example for a UDF definition:

Reviewer: I think it'd be great to add some more explanation here (some of which could be before or after the code example), i.e.:

* What does this UDF do?
* Are there multiple ways to define/register UDFs?
* Why do we use "Column" on the left but "string" on the right?
* Some info that the first "string" is the type of the input and the second "string" is the type of the output

Author: Have added some explanation and link to relevant sections in our code for easier understanding. Please let me know if this looks fine, thanks!

Reviewer: Looks great, minor comment :)

```csharp
string s1 = "hello";
Func<Column, Column> udf = Udf<string, string>(
    str => $"{s1} {str}");

```

And for a sample Dataframe, let's take the following Dataframe `df`:

```text
    +-------+
    |   name|
    +-------+
    |Michael|
    |   Andy|
    | Justin|
    +-------+
```

Now let's apply the above defined `udf` to the dataframe `df`:

```csharp
DataFrame udfResult = df.Select(udf(df["name"]));
```

This would return the below as the Dataframe `udfResult`:

```text
    +-------------+
    |         name|
    +-------------+
    |hello Michael|
    |   hello Andy|
    | hello Justin|
    +-------------+
```

## Good to know while implementing UDFs

> One behavior to be aware of while implementing UDFs in .NET for Apache Spark is the way target of the UDF gets serialized. .NET for Apache Spark uses .NET Core which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes the target of reflection for serialization. Let's take an example to illustrate what that means:

Reviewer: ```suggestion
One behavior to be aware of while implementing UDFs in .NET for Apache Spark is how the target of the UDF gets serialized. .NET for Apache Spark uses .NET Core, which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes the target of reflection for serialization. Let's take an example to illustrate what that means.
```

Reviewer: Could we explain delegates, reflection, and shared closure?

Author: Is it better to provide links to official docs for these terms or do we want a 1-2 line description in this doc?

> Sample user code:

Reviewer: Rather than "Sample user code," perhaps we could have a sentence or two explaining what the following user code does.

Author: Done, thanks!

    }
}
```

The above C# code generates the following IL (Intermediate language) code from the compiler:

```csharp
public class C
{
    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_0
    {
        public string s1;

        public string s2;

        internal string <M>b__0(string str)
        {
            return s1;
        }

        internal string <M>b__1(string str)
        {
            return s2;
        }
    }

    public void M()
    {
        <>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
        <>c__DisplayClass0_.s1 = "s1";
        <>c__DisplayClass0_.s2 = "s2";
        Func<string, string> func = new Func<string, string>(<>c__DisplayClass0_.<M>b__0);
        Func<string, string> func2 = new Func<string, string>(<>c__DisplayClass0_.<M>b__1);
    }
}
```
As can be seen in the above IL code, both `func` and `func2` share the same closure `<>c__DisplayClass0_0`, which is the target that is serialized when serializing the delegates `func` and `func2`. Hence, even though `Func<string, string> a` is only referencing `s1`, `s2` also gets serialized when sending over the bytes to the workers.

> This can lead to some unexpected behaviors at runtime (for example in the case of using Broadcast variables, as explained in more detail in [this guide](broadcast-guide.md)), which is why we recommend restricting the visibility of the variables used in a function to that function's scope.

Reviewer: ```suggestion

This can lead to some unexpected behaviors at runtime (like in the case of using [broadcast variables](broadcast-guide.md)), which is why we recommend restricting the visibility of the variables used in a function to that function's scope.
```

Reviewer: Added extra space before sentence so that it starts as a new paragraph (if you think that'd be okay).

Taking the above example to better explain what that means:

Recommended user code:

> ```csharp

Reviewer: Is this the same recommended user code as above? If so, rather than repeating it (which will make the doc longer and could be a bit confusing), I think it'd make sense just to state that you're using the same user code to accomplish task X, as was shown above.

Author: No actually this is not supposed to be the same code (my bad), thanks for pointing that out! Will make the change.

using System;

public class C {
    public void M() {
        string s1 = "s1";
        string s2 = "s2";
        Func<string, string> a = str => s1;
        Func<string, string> b = str => s2;
    }
}
```

> The above C# code generates the following IL (Intermediate language) code from the compiler:

Reviewer 2: Technically speaking, the following csharp code is not an IL, but C# disassembly (I think). You can select to see IL from sharplab.io.

Author: Yes you are right, will change the wording here.

```csharp

> public class C

Reviewer 2: Let's give a credit to sharplap.io

{
    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_0
    {
        public string s1;

        internal string <M>b__0(string str)
        {
            return s1;
        }
    }

    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_1
    {
        public string s2;

        internal string <M>b__1(string str)
        {
            return s2;
        }
    }

    public void M()
    {
        <>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
        <>c__DisplayClass0_.s1 = "s1";
        Func<string, string> func = new Func<string, string>(<>c__DisplayClass0_.<M>b__0);
        <>c__DisplayClass0_1 <>c__DisplayClass0_2 = new <>c__DisplayClass0_1();
        <>c__DisplayClass0_2.s2 = "s2";
        Func<string, string> func2 = new Func<string, string>(<>c__DisplayClass0_2.<M>b__1);
    }
}
```

> Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively, which when used as target for serialization mean that no more than the referenced variables get serialized for the delegate.

Reviewer: ```suggestion
Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively. When used as the target for serialization, nothing other than the referenced variables will get serialized for the delegate.
```

> 

Reviewer: Perhaps include a section with a couple sentences of wrap up, and/or link to other relevant articles and coding examples?

Author: Added a summary with links to articles, please let me know if this looks ok. Thanks!

Reviewer: Looks really good! Added minor wording suggestion.

# Guide to User-Defined Functions (UDFs)

This is a guide to show how to use UDFs in .NET for Apache Spark.

## What are UDFs

[User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the system's built-in functionality. They transform values from a single row within a table to produce a single corresponding output value per row based on the logic defined in the UDF.

Let's take the following as an example for a UDF definition:

```csharp
    string s1 = "hello";
    Func<Column, Column> udf = Udf<string, string>(
        str => $"{s1} {str}");

```

> The above defined UDF takes a `string` as an input (in the form of a [Column](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/Column.cs#L14) of a [Dataframe](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/DataFrame.cs#L24)), and returns a `string` with `hello` appended infront of the input.

Reviewer: nit
```suggestion
The above defined UDF takes a `string` as an input (in the form of a [Column](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/Column.cs#L14) of a [Dataframe](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/DataFrame.cs#L24)), and returns a `string` with `hello` appended in front of the input.
```

        {
            public string s1;

            internal string <M>b__0(string str)
            {
                return s1;
            }
        }

        [CompilerGenerated]
        private sealed class <>c__DisplayClass0_1
        {
            public string s2;

            internal string <M>b__1(string str)
            {
                return s2;
            }
        }

        public void M()
        {
            <>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
            <>c__DisplayClass0_.s1 = "s1";
            Func<string, string> func = new Func<string, string>(<>c__DisplayClass0_.<M>b__0);
            <>c__DisplayClass0_1 <>c__DisplayClass0_2 = new <>c__DisplayClass0_1();
            <>c__DisplayClass0_2.s2 = "s2";
            Func<string, string> func2 = new Func<string, string>(<>c__DisplayClass0_2.<M>b__1);
        }
    }
```

Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively. When used as the target for serialization, nothing other than the referenced variables will get serialized for the delegate.

> This above behavior is important to keep in mind while implementing multiple UDFs in a common scope. To know more about UDFs in general please go through the following articles that explain UDFs and how to use them: [UDFs in databricks(scala)](https://docs.databricks.com/spark/latest/spark-sql/udf-scala.html), [Spark UDFs and some gotchas](https://medium.com/@achilleus/spark-udfs-we-can-use-them-but-should-we-use-them-2c5a561fde6d).

Reviewer: nit wording

```suggestion
This above behavior is important to keep in mind while implementing multiple UDFs in a common scope. 

To learn more about UDFs in general, please review the following articles that explain UDFs and how to use them: [UDFs in databricks(scala)](https://docs.databricks.com/spark/latest/spark-sql/udf-scala.html), [Spark UDFs and some gotchas](https://medium.com/@achilleus/spark-udfs-we-can-use-them-but-should-we-use-them-2c5a561fde6d).
```

# Guide to User-Defined Functions (UDFs)

This is a guide to show how to use UDFs in .NET for Apache Spark.

## What are UDFs

[User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the system's built-in functionality. They transform values from a single row within a table to produce a single corresponding output value per row based on the logic defined in the UDF.

Let's take the following as an example for a UDF definition:

```csharp

>     string s1 = "hello";

Reviewer 2: for all examples in this PR, can you remove leading identations?

For a sample Dataframe, let's take the following Dataframe `df`:

```text
    +-------+
    |   name|
    +-------+
    |Michael|
    |   Andy|
    | Justin|
    +-------+
```

Now let's apply the above defined `udf` to the dataframe `df`:

```csharp
    DataFrame udfResult = df.Select(udf(df["name"]));
```

This would return the below as the Dataframe `udfResult`:

```text
    +-------------+
    |         name|
    +-------------+
    |hello Michael|
    |   hello Andy|
    | hello Justin|
    +-------------+
```
To get a better understanding of how to implement UDFs, please take a look at the [UDF helper functions](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/Functions.cs#L3616) and some [test examples](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs#L49).

## Good to know while implementing UDFs

> One behavior to be aware of while implementing UDFs in .NET for Apache Spark is how the target of the UDF gets serialized. .NET for Apache Spark uses .NET Core, which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes the target of reflection for serialization. Let's take an example to illustrate what that means.

Reviewer 2: I think we need to explain what `target` is and the link to target in msdn doc.

Reviewer 2: I think if you are explaining serializing, you may need to explain a bit more. This sounds like we are just serializing the target. Also, please have a link to the code where the serialization happens. Maybe a section just dedicated to a serialization would help, I think.

```csharp
using System;

public class C {
    public void M() {
        string s1 = "s1";
        string s2 = "s2";
        Func<string, string> a = str => s1;
        Func<string, string> b = str => s2;
    }
}
```

The above C# code generates the following C# disassembly (credit source: [sharplab.io](sharplab.io)) code from the compiler:

```csharp
public class C
{
    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_0
    {
        public string s1;

        public string s2;

        internal string <M>b__0(string str)
        {
            return s1;
        }

        internal string <M>b__1(string str)
        {
            return s2;
        }
    }

    public void M()
    {
        <>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
        <>c__DisplayClass0_.s1 = "s1";
        <>c__DisplayClass0_.s2 = "s2";
        Func<string, string> func = new Func<string, string>(<>c__DisplayClass0_.<M>b__0);
        Func<string, string> func2 = new Func<string, string>(<>c__DisplayClass0_.<M>b__1);
    }
}
```

> As can be seen in the above IL code, both `func` and `func2` share the same closure `<>c__DisplayClass0_0`, which is the target that is serialized when serializing the delegates `func` and `func2`. Hence, even though `Func<string, string> a` is only referencing `s1`, `s2` also gets serialized when sending over the bytes to the workers.

Author: You're right, it is the decompiled C# code, will make the change
